onnxruntime/python/tools/transformers/convert_generation.py

```diff
@@ -1066,6 +1075,40 @@ def update_decoder_subgraph_past_present_share_buffer(subg):
     return subg
 
 
+def update_decoder_subgraph_use_decoder_masked_multihead_attention(subg) -> bool:
```

Comment:
Another solution is to change fusion script to fuse DecoderMaskedMultiheadAttention instead of Attention. That's the reason we added an option in export:
https://github.com/microsoft/onnxruntime/blob/f88b97ede272e54d67e7f4da165359df4b534a4a/onnxruntime/python/tools/transformers/models/gpt2/convert_to_onnx.py#L164
Then in fusion, we can set an option in [FusionOptions](https://github.com/microsoft/onnxruntime/blob/f88b97ede272e54d67e7f4da165359df4b534a4a/onnxruntime/python/tools/transformers/fusion_options.py#L22), then the fusion could create DecoderMaskedMultiheadAttention or Attention accordingly based on the option:
https://github.com/microsoft/onnxruntime/blob/f88b97ede272e54d67e7f4da165359df4b534a4a/onnxruntime/python/tools/transformers/fusion_gpt_attention.py#L193
The benefits of such approach:
(1) Less coupling with Attention op. 
(2) make it easier to test since there are tools for GPT-2 model, and we can test without convert generation.

It is fine that we use update graph approach here. Just a suggestion that we might change it in the future.

Reply:
Thanks for this - I agree- this seems like the right long-term solution and gives a way to generate the "decoder" and "init_decoder" subgraphs using the `convert_to_onnx.py` entry-point by providing the `stage` parameter.
However, the way we generate the "init_decoder" subgraph today in `convert_generation.py` is to use the optimized GPT2 subgraph from `convert_to_onnx.py` (using the default stage = 0 option) and then modifies the subgraph to include the "Slice" nodes to generate the "init_decoder" subgraph.
So, to incorporate your suggestion, we will have to move the "Slice" node insertion logic into `convert_to_onnx` and execute it when stage = 2 and requires appropriate wiring changes in `convert_generation.py`. Perhaps, we could defer this for a future PR ?